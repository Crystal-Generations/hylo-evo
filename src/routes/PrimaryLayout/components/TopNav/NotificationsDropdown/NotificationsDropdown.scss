.header-content {
  composes: header-content from '../TopNavDropdown/TopNavDropdown.scss';
}

.tab {
  composes: header-link from '../TopNavDropdown/TopNavDropdown.scss';
  margin-right: 15px;
}

.active {
  color: $color-rhino;
  border-bottom: 1px solid $color-rhino;
}

.mark-read {
  composes: header-link from '../TopNavDropdown/TopNavDropdown.scss';
  margin-left: auto;
}

.threads {
  composes: body from '../TopNavDropdown/TopNavDropdown.scss';
}

.thread {
  display: flex;
  align-items: start;
  cursor: pointer;
  &:first-child {
    margin-top: 5px;
  }
  &:last-child {
    .message-content {
      border: none;
    }
  }
}

.unread {
  background-color: $color-saltpan;
}

.image-wraper {
  width: $space-10x;
  margin: 0 10px 0 15px;
  padding-top: 20px;
}

.message-content {
  display: flex;
  flex-direction: column;
  align-items: start;
  padding: 20px 15px 15px 0;
  border-bottom: 1px solid $color-black-squeeze;
}

.name {
  composes: username-cards from 'css/typography.scss';
  margin-bottom: 10px;
}

.body {
  composes: caption-lt-lg from 'css/typography.scss';
}

.date {
  composes: caption-lt-lg from 'css/typography.scss';
  color: $color-slate-gray-80;
  font-size: 12px;
}
