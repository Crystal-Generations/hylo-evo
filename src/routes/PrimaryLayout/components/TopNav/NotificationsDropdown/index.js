import { HOLOCHAIN_ACTIVE } from 'util/holochain'
import component from './NotificationsDropdown'
import holochainConnector from './NotificationsDropdown.holochain.connector'
import connector from './NotificationsDropdown.connector'

export default HOLOCHAIN_ACTIVE
  ? holochainConnector(component)
  : connector(component)
