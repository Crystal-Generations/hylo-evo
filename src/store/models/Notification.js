import { attr, fk, Model } from 'redux-orm'

const Notification = Model.createClass({
  toString () {
    return `Message: ${this.id}`
  }
})

export default Notification

Notification.modelName = 'Notification'

Notification.fields = {
  id: attr(),
  action: attr(),
  activity: fk('Activity'),
  meta: attr()
}

export const ACTION_NEW_COMMENT = 'newComment'
export const ACTION_TAG = 'tag'
