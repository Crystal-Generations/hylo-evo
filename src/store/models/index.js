import { ORM } from 'redux-orm'
import './Model.extension'
import Comment from './Comment'
import Community, { CommunityModerator } from './Community'
import FeedItem from './FeedItem'
import Me from './Me'
import Membership from './Membership'
import Person from './Person'
import PersonConnection from './PersonConnection'
import Message from './Message'
import MessageThread from './MessageThread'
import Post, { PostFollower, PostCommenter } from './Post'
import Topic from './Topic'
import TopicSubscription from './TopicSubscription'
import Vote from './Vote'

export const orm = new ORM()
orm.register(
  Comment,
  Community,
  CommunityModerator,
  FeedItem,
  Me,
  Membership,
  Message,
  MessageThread,
  Person,
  PersonConnection,
  Post,
  PostFollower,
  PostCommenter,
  Topic,
  TopicSubscription,
  Vote
)

export default orm
