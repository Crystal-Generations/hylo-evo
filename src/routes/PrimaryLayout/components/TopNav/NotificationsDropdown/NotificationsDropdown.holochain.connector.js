import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import getMe from 'store/selectors/getMe'
import { urlForNotification } from 'store/models/Notification'

// * Not used by Holochain
export function mapStateToProps (state, props) {
  return {
    currentUser: getMe(state, props),
    notifications: [],
    fetchNotifications: () => {},
    markActivityRead: () => {},
    markAllActivitiesRead: () => {}
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    goToNotification: notification => dispatch(push(urlForNotification(notification)))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
