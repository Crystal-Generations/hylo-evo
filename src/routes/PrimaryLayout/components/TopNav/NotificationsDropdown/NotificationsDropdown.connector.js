import { connect } from 'react-redux'
import { fetchNotifications } from './NotificationsDropdown.store.js'
import { getMe } from 'store/selectors/getMe'
import faker from 'faker'
import samplePost, { fakePerson } from 'components/PostCard/samplePost'
import { ACTION_NEW_COMMENT, ACTION_TAG } from 'store/models/Notification'

faker.seed(1235)

const NOTIFICATIONS = [
  {
    id: 1,
    actor: fakePerson(),
    action: ACTION_NEW_COMMENT,
    meta: {},
    post: {...samplePost(), title: 'Our Oceans'},
    comment: {
      text: "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast"
    },
    unread: true,
    createdAt: faker.date.recent().toString()
  },
  {
    id: 2,
    actor: fakePerson(),
    action: ACTION_TAG,
    post: {...samplePost(), title: 'Our Oceans'},
    meta: {reasons: ['tag: request']},
    unread: true,
    createdAt: faker.date.recent().toString()
  },
  {
    id: 3,
    actor: fakePerson(),
    action: ACTION_NEW_COMMENT,
    post: {...samplePost(), title: 'Our Oceans'},
    meta: {},
    comment: {
      text: "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast"
    },
    unread: false,
    createdAt: faker.date.recent().toString()
  },
  {
    id: 4,
    actor: fakePerson(),
    action: ACTION_TAG,
    post: {...samplePost(), title: 'Our Oceans'},
    meta: {reasons: ['tag: request']},
    unread: false,
    createdAt: faker.date.recent().toString()
  }
]

export function mapStateToProps (state, props) {
  return {
    // currentUser: getMe(state, props), ???
    notifications: NOTIFICATIONS
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    fetchNotifications: () => dispatch(fetchNotifications()),
    goToNotification: notification => () => console.log('going to notification', notification),
    markAsRead: () => console.log('mark as read')
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
