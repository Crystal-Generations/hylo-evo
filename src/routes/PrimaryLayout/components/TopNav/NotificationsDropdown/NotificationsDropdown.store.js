import { FETCH_NOTIFICATIONS } from 'store/constants'
import { createSelector as ormCreateSelector } from 'redux-orm'
import orm from 'store/models'

export function fetchNotifications () {
  return {
    type: FETCH_NOTIFICATIONS,
    graphql: {
      query: `{
        me {
          name
          notifications (first: 10, order: "desc") {
            total
            hasMore
            items {
              id
              createdAt
              activity {
                id
                actor {
                  id
                  name
                  avatarUrl
                }
                comment {
                  id
                  text
                }
                post {
                  id
                  title
                }
                community {
                  id
                  name
                  slug
                }
                meta {
                  reasons
                }
                action
                unread
              }
            }
          }
        }
      }`
    },
    meta: {
      extractModel: 'Me'
    }
  }
}

export const getNotifications = ormCreateSelector(
  orm,
  state => state.orm,
  (session) => {
    return session.Notification.all().toModelArray()
  }
)
