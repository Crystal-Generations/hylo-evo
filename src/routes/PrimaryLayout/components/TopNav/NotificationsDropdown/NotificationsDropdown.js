import React, { PropTypes, Component } from 'react'
import './NotificationsDropdown.scss'
const { object, array, string, func } = PropTypes
import { humanDate, textLength, truncate } from 'hylo-utils/text'
import cx from 'classnames'
import RoundImageRow from 'components/RoundImageRow'
import TopNavDropdown from '../TopNavDropdown'

export default class NotificationsDropdown extends Component {
  static propTypes = {
    fetchThreads: func,
    toggleChildren: object,
    threads: array,
    className: string,
    goToThread: func
  }

  constructor (props) {
    super(props)
    this.state = {
      showingUnread: false
    }
  }

  componentDidMount () {
    this.props.fetchThreads()
  }

  render () {
    const {
      toggleChildren, threads, className, goToThread, currentUser, markAsRead
    } = this.props
    const { showingUnread } = this.state

    const showRecent = () => this.setState({showingUnread: false})
    const showUnread = () => this.setState({showingUnread: true})

    return <TopNavDropdown
      className={className}
      toggleChildren={toggleChildren}
      header={
        <div styleName='header-content'>
          <span onClick={showRecent} styleName={cx('tab', {active: !showingUnread})}>
            Recent
          </span>
          <span onClick={showUnread} styleName={cx('tab', {active: showingUnread})}>
            Unread
          </span>
          <span onClick={markAsRead} styleName='mark-read'>Mark all as read</span>
        </div>}
      body={
        <div styleName='threads'>
          {threads.map(thread => <Thread
            thread={thread}
            goToThread={goToThread}
            currentUserId={currentUser.id}
            key={thread.id} />)}
        </div>
      } />
  }
}

const participantNames = participants => {
  const length = participants.length
  if (length === 1) {
    return participants[0].name
  } else if (length === 2) {
    return `${participants[0].name} and ${participants[1].name}`
  } else if (length > 2) {
    const n = length - 2
    return `${participants[0].name}, ${participants[1].name} and ${n} other${n > 1 ? 's' : ''}`
  }
}

export function Thread ({ thread, goToThread, currentUserId }) {
  const message = thread.messages[0]
  if (!message || !message.text) return null
  const participants = thread.participants.filter(p => p.id !== currentUserId)

  const unread = thread.lastReadAt < thread.updatedAt

  var { text } = message
  if (message.creator.id === currentUserId) {
    text = `You: ${text}`
  }

  const maxMessageLength = 145

  if (textLength(text) > maxMessageLength) {
    text = `${truncate(text, maxMessageLength)}...`
  }

  return <li styleName={cx('thread', {unread})}
    onClick={goToThread(thread.id)}>
    <div styleName='image-wraper'>
      <RoundImageRow imageUrls={participants.map(p => p.avatarUrl)} vertical ascending cap='2' />
    </div>
    <div styleName='message-content'>
      <div styleName='name'>{participantNames(participants)}</div>
      <div styleName='body'>{text}</div>
      <div styleName='date'>{humanDate(thread.updatedAt)}</div>
    </div>
  </li>
}
