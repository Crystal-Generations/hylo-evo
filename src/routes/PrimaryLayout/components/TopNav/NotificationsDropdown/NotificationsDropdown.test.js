import NotificationsDropdown, { Notification, NotificationHeader, NotificationBody } from './NotificationsDropdown'
import { shallow } from 'enzyme'
import React from 'react'
import { ACTION_NEW_COMMENT, ACTION_TAG } from 'store/models/Notification'

const u1 = {id: 1, name: 'Charles Darwin', avatarUrl: 'foo.png'}
const u2 = {id: 2, name: 'Marie Curie', avatarUrl: 'bar.png'}
const u3 = {id: 3, name: 'Arthur Fonzarelli', avatarUrl: 'baz.png'}

const commentNotification = {
  id: 1,
  actor: u2,
  action: ACTION_NEW_COMMENT,
  meta: {},
  post: {title: 'Our Oceans'},
  comment: {
    text: "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast"
  },
  unread: true,
  createdAt: new Date('1995-12-17T03:23:00')
}

const tagNotification = {
  id: 2,
  actor: u3,
  action: ACTION_TAG,
  post: {title: 'I have so many things I need!'},
  meta: {reasons: ['tag: request']},
  unread: true,
  createdAt: new Date('1995-12-17T03:23:00')
}

const notifications = [
  commentNotification,
  tagNotification,
  {...commentNotification, unread: false},
  {...tagNotification, unread: false}
]

describe('NotificationsDropdown', () => {
  it('renders correctly with an empty list', () => {
    const wrapper = shallow(<NotificationsDropdown notifications={[]} currentUser={u1} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly with a list of threads', () => {
    const wrapper = shallow(<NotificationsDropdown notifications={notifications} currentUser={u1} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('Notification', () => {
  it('renders correctly with a comment notification', () => {
    const wrapper = shallow(<Notification notification={commentNotification} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly with a tag notification', () => {
    const wrapper = shallow(<Notification notification={tagNotification} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('NotificationHeader', () => {
  it('renders correctly with a comment notification', () => {
    const wrapper = shallow(<NotificationHeader notification={commentNotification} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly with a tag notification', () => {
    const wrapper = shallow(<NotificationHeader notification={tagNotification} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('NotificationBody', () => {
  it('renders correctly with a comment notification', () => {
    const wrapper = shallow(<NotificationBody notification={commentNotification} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly with a tag notification', () => {
    const wrapper = shallow(<NotificationBody notification={tagNotification} />)
    expect(wrapper).toMatchSnapshot()
  })
})
