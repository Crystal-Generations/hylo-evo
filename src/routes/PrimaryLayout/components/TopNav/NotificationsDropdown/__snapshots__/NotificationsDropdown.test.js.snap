// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders correctly with a comment notification 1`] = `
<li
  data-styleName="notification unread"
  onClick={[Function]}
>
  <div
    data-styleName="image-wraper"
  >
    <RoundImage
      url="bar.png"
    />
  </div>
  <div
    data-styleName="content"
  >
    <NotificationHeader
      notification={
        Object {
          "action": "newComment",
          "actor": Object {
            "avatarUrl": "bar.png",
            "id": 2,
            "name": "Marie Curie",
          },
          "comment": Object {
            "text": "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast",
          },
          "createdAt": 1995-12-17T03:23:00.000Z,
          "id": 1,
          "meta": Object {},
          "post": Object {
            "title": "Our Oceans",
          },
          "unread": true,
        }
      }
    />
    <NotificationBody
      notification={
        Object {
          "action": "newComment",
          "actor": Object {
            "avatarUrl": "bar.png",
            "id": 2,
            "name": "Marie Curie",
          },
          "comment": Object {
            "text": "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast",
          },
          "createdAt": 1995-12-17T03:23:00.000Z,
          "id": 1,
          "meta": Object {},
          "post": Object {
            "title": "Our Oceans",
          },
          "unread": true,
        }
      }
    />
    <div
      data-styleName="date"
    >
      21 years ago
    </div>
  </div>
</li>
`;

exports[`Notification renders correctly with a tag notification 1`] = `
<li
  data-styleName="notification unread"
  onClick={[Function]}
>
  <div
    data-styleName="image-wraper"
  >
    <RoundImage
      url="baz.png"
    />
  </div>
  <div
    data-styleName="content"
  >
    <NotificationHeader
      notification={
        Object {
          "action": "tag",
          "actor": Object {
            "avatarUrl": "baz.png",
            "id": 3,
            "name": "Arthur Fonzarelli",
          },
          "createdAt": 1995-12-17T03:23:00.000Z,
          "id": 2,
          "meta": Object {
            "reasons": Array [
              "tag: request",
            ],
          },
          "post": Object {
            "title": "I have so many things I need!",
          },
          "unread": true,
        }
      }
    />
    <NotificationBody
      notification={
        Object {
          "action": "tag",
          "actor": Object {
            "avatarUrl": "baz.png",
            "id": 3,
            "name": "Arthur Fonzarelli",
          },
          "createdAt": 1995-12-17T03:23:00.000Z,
          "id": 2,
          "meta": Object {
            "reasons": Array [
              "tag: request",
            ],
          },
          "post": Object {
            "title": "I have so many things I need!",
          },
          "unread": true,
        }
      }
    />
    <div
      data-styleName="date"
    >
      21 years ago
    </div>
  </div>
</li>
`;

exports[`NotificationBody renders correctly with a comment notification 1`] = `
<div
  data-styleName="body"
>
  <span
    data-styleName="bold"
  >
    Marie
  </span>
   Wrote: "
  I live right next to there and can come help out. I've never done â€¦
  "
</div>
`;

exports[`NotificationBody renders correctly with a tag notification 1`] = `
<div
  data-styleName="body"
>
  <span
    data-styleName="bold"
  >
    Arthur
  </span>
   Wrote: "
  I have so many things I need!
  "
</div>
`;

exports[`NotificationHeader renders correctly with a comment notification 1`] = `
<div
  data-styleName="header"
>
  New Comment on 
  <span
    data-styleName="bold"
  >
    Our Oceans
  </span>
</div>
`;

exports[`NotificationHeader renders correctly with a tag notification 1`] = `
<div
  data-styleName="header"
>
  New Post in 
  <span
    data-styleName="bold"
  >
    #
    request
  </span>
</div>
`;

exports[`NotificationsDropdown renders correctly with a list of threads 1`] = `
<Connect(TopNavDropdown)
  body={
    <div
      data-styleName="notifications"
    >
      <Notification
        notification={
          Object {
            "action": "newComment",
            "actor": Object {
              "avatarUrl": "bar.png",
              "id": 2,
              "name": "Marie Curie",
            },
            "comment": Object {
              "text": "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast",
            },
            "createdAt": 1995-12-17T03:23:00.000Z,
            "id": 1,
            "meta": Object {},
            "post": Object {
              "title": "Our Oceans",
            },
            "unread": true,
          }
        }
      />
      <Notification
        notification={
          Object {
            "action": "tag",
            "actor": Object {
              "avatarUrl": "baz.png",
              "id": 3,
              "name": "Arthur Fonzarelli",
            },
            "createdAt": 1995-12-17T03:23:00.000Z,
            "id": 2,
            "meta": Object {
              "reasons": Array [
                "tag: request",
              ],
            },
            "post": Object {
              "title": "I have so many things I need!",
            },
            "unread": true,
          }
        }
      />
      <Notification
        notification={
          Object {
            "action": "newComment",
            "actor": Object {
              "avatarUrl": "bar.png",
              "id": 2,
              "name": "Marie Curie",
            },
            "comment": Object {
              "text": "I live right next to there and can come help out. I've never done petitioning but I'm sure it's an absolute blast",
            },
            "createdAt": 1995-12-17T03:23:00.000Z,
            "id": 1,
            "meta": Object {},
            "post": Object {
              "title": "Our Oceans",
            },
            "unread": false,
          }
        }
      />
      <Notification
        notification={
          Object {
            "action": "tag",
            "actor": Object {
              "avatarUrl": "baz.png",
              "id": 3,
              "name": "Arthur Fonzarelli",
            },
            "createdAt": 1995-12-17T03:23:00.000Z,
            "id": 2,
            "meta": Object {
              "reasons": Array [
                "tag: request",
              ],
            },
            "post": Object {
              "title": "I have so many things I need!",
            },
            "unread": false,
          }
        }
      />
    </div>
  }
  header={
    <div
      data-styleName="header-content"
    >
      <span
        data-styleName="tab active"
        onClick={[Function]}
      >
        Recent
      </span>
      <span
        data-styleName="tab"
        onClick={[Function]}
      >
        Unread
      </span>
      <span
        data-styleName="mark-read"
      >
        Mark all as read
      </span>
    </div>
  }
/>
`;

exports[`NotificationsDropdown renders correctly with an empty list 1`] = `
<Connect(TopNavDropdown)
  body={
    <div
      data-styleName="notifications"
    />
  }
  header={
    <div
      data-styleName="header-content"
    >
      <span
        data-styleName="tab active"
        onClick={[Function]}
      >
        Recent
      </span>
      <span
        data-styleName="tab"
        onClick={[Function]}
      >
        Unread
      </span>
      <span
        data-styleName="mark-read"
      >
        Mark all as read
      </span>
    </div>
  }
/>
`;
